// This proto describes the beats information attached to a track.
syntax = "proto2";

package mixxx.track.io;

option optimize_for = LITE_RUNTIME;

enum Source {
  ANALYZER = 0;
  FILE_METADATA = 1;
  USER = 2;
}

enum Type {
  BEAT = 0;
  BAR = 1;
  PHRASE = 2;
}

message TimeSignature {
  optional int32 beats_per_bar = 1 [ default = 4];
  optional int32 note_value = 2 [ default = 4 ];
}

message Bpm {
  optional double bpm = 1;
  optional Source source = 2 [ default = ANALYZER ];
}

message Beat {
  optional double frame_position = 1 [ default = 0 ];
  optional bool enabled = 2 [ default = true ];
  optional Source source = 3 [ default = ANALYZER ];
  optional Type type = 4 [ default = BEAT ];
  optional TimeSignature signature = 5;
}

message Beats {
  repeated Beat beat = 1;
}

// These are legacy data structures for migrating between Mixxx versions
message LegacyBeat {
  optional int32 frame_position = 1;
  optional bool enabled = 2 [ default = true ];
  optional Source source = 3 [ default = ANALYZER ];
}

message LegacyBeatMap {
  repeated LegacyBeat beat = 1;
}

message LegacyBeatGrid {
  optional Bpm bpm = 1;
  optional LegacyBeat first_beat = 2;
}